packages/plugins/experimental/plugin-wnfs/src/extensions/image.ts

```diff
@@ -0,0 +1,247 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { syntaxTree } from '@codemirror/language';
+import { type EditorState, type Extension, StateField, type Range, StateEffect } from '@codemirror/state';
+import { Decoration, EditorView, WidgetType, ViewPlugin, type ViewUpdate } from '@codemirror/view';
+import type { Blockstore } from 'interface-blockstore';
+import type { PrivateDirectory, PrivateForest } from 'wnfs';
+
+import { type Space } from '@dxos/react-client/echo';
+
+import { store } from '../common';
+import { loadWnfs } from '../load';
+
+export type ImageOptions = {
+  blockstore: Blockstore;
+  space: Space;
+};
+
+export const image = (options: ImageOptions): Extension[] => {
+  return [stateField(options), viewPlugin(options)];
+};
+
+// GLOBAL STATE
+
+/** Indexed by space id */
+const loadedWnfsInstances: Record<string, { directory: PrivateDirectory; forest: PrivateForest }> = {};
+
+// EFFECT + EXTENSIONS
+
+const stateEffect = StateEffect.define<{
+  decorations: Range<Decoration>[];
+  from: number;
+  to: number;
+}>({});
+
+const stateField = (options: ImageOptions) =>
+  StateField.define({
+    create: () => {
+      return Decoration.none;
+    },
+    update: (value, tr) => {
+      let eff;
+      for (const effect of tr.effects) {
+        if (effect.is(stateEffect)) {
+          eff = effect.value;
+        }
+      }
+      if (eff === undefined) {
+        return value;
+      }
+      return value.map(tr.changes).update({
```

Comment:
i think this would be clearly (and more compact) if you just put this inline inside the for/if loop.